streaming/src/event_server.h

```diff
@@ -0,0 +1,216 @@
+      urgent_buffer_.pop();
+    } else {
+      buffer_.pop();
+    }
+    full_cv_.notify_all();
+  }
+
+  bool Get(Event &evt) {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    while (Empty() && is_started_) {
+      empty_cv_.wait(lock);
+    }
+    if (!is_started_) {
+      return false;
+    }
+    if (!urgent_buffer_.empty()) {
+      urgent_ = true;
+      evt = urgent_buffer_.front();
+    } else {
+      urgent_ = false;
+      evt = buffer_.front();
+    }
+    return true;
+  }
+
+  Event PopAndGet() {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    while (Empty() && is_started_) {
+      empty_cv_.wait(lock);
+    }
+    if (!is_started_) {
+      return Event();
+    }
+    if (!urgent_buffer_.empty()) {
+      Event res = urgent_buffer_.front();
+      urgent_buffer_.pop();
+      if (Full()) {
+        full_cv_.notify_one();
+      }
+      return res;
+    }
+    Event res = buffer_.front();
+    buffer_.pop();
+    if (Size() + 1 == capacity_) full_cv_.notify_one();
+    return res;
+  }
+
+  Event &Front() {
+    std::unique_lock<std::mutex> lock(ring_buffer_mutex_);
+    if (urgent_buffer_.size()) {
+      return urgent_buffer_.front();
+    }
+    return buffer_.front();
+  }
+
+  inline bool Empty() const { return buffer_.empty() && urgent_buffer_.empty(); }
+
+  inline bool Full() const { return buffer_.size() + urgent_buffer_.size() == capacity_; }
+
+  inline size_t Size() const { return buffer_.size() + urgent_buffer_.size(); }
+
+  inline size_t UrgentBufferSize() const { return urgent_buffer_.size(); }
+
+  inline size_t Capacity() const { return capacity_; }
+
+ private:
+  std::mutex ring_buffer_mutex_;
+  std::condition_variable empty_cv_;
+  std::condition_variable full_cv_;
+  // Normal events wil be pushed into buffer_.
+  std::queue<Event> buffer_;
+  // This field urgent_buffer_ is used for serving urgent event.
+  std::queue<Event> urgent_buffer_;
+  // Urgent event will be poped out first if urgent_ flag is true.
+  bool urgent_;
+  size_t capacity_;
+  bool is_started_;
+};
+
+class EventServer {
+ public:
+  /// User-define event handle for different types.
+  typedef std::function<bool(ProducerChannelInfo *info)> Handle;
+
+  EventServer(uint32_t event_size = kEventQueueCapacity);
+
+  ~EventServer();
+
+  void Run();
+
+  void Stop();
+
+  bool Register(const EventType &type, const Handle &handle);
+
+  void Push(const Event &event);
+
+  /// A single thread should be invoked to run this loop function, so that
+  /// event server can poll and execute registered callback function event
+  /// one by one.
+  void LoopThreadHandler();
+
+  inline size_t EventNums() const { return event_queue_->Size(); }
+
+  void RemoveDestroyedChannelEvent(const std::vector<ObjectID> &removed_ids);
+
+ private:
+  void Execute(Event &event);
+
+ private:
+  std::unordered_map<EventType, Handle, EnumTypeHash> event_handle_map_;
+  std::shared_ptr<EventQueue> event_queue_;
```

Comment:
Why we need this event queue? I think we can use two blocking queue instead. One for`urgent_buffer`, one for normal `buffer`. And this logic is still encapsulated in `event_server`, not leaded to user code.

Reply:
You mean we can control all of event in eventserver.
I consider that event queue can handle different prioprity events inner so that event server do not care how many event types or event prioprities we have, which should be better than encapsualting a lot of blocking queue in event server. 
If not, we should change event server to `PriorityEventServer`.
Finally, I think EventQueue can reduce complexity.